File: ruby/easy/valid_word_abbreviation.rb
# @param {String} word
# @param {String} abbr
# @return {Boolean}
def valid_word_abbreviation(word, abbr)
    i,j = 0,0
    abbr_counter,num = 0,0
    while i < word.length && j < abbr.length
        if is_numeric(abbr[j])
            num = (num * 10) + get_number(abbr[j])
            return false if num == 0
            j += 1 and next
        end
        abbr_counter += num
        while(num > 0 && i < word.length)
            num -= 1
            i += 1
        end
        return false if i >= word.length || word[i] != abbr[j]
        i += 1 and j += 1 and abbr_counter += 1
    end
    return ((i + num) == word.length) && (j == abbr.length) && ((abbr_counter + num) == word.length)
end

def is_numeric(c)
    return c.ord >= '0'.ord && c.ord <= '9'.ord
end

def get_number(c)
    return c.ord - '0'.ord
end

File: javascript/easy/validWordAbbreviation.js
/**
 * @param {string} word
 * @param {string} abbr
 * @return {boolean}
 */
var validWordAbbreviation = function(word, abbr) {
    let i = 0, j = 0, num = 0, abbrCounter = 0;
    while(i < word.length && j < abbr.length) {
        if(isNumeric(abbr[j])) {
            num = (num * 10) + getNumber(abbr[j]);
            if(num === 0) {
                return false;
            }
            j++; continue;
        }
        
        abbrCounter += num;
        while(num > 0 && i < word.length) {
            num--; i++;
        }
        
        if(i >= word.length || word[i] !== abbr[j]) {
            break;
        }
        i++; j++;
        abbrCounter++;
    }
    return ((i+num) === word.length) && (j === abbr.length) && ((abbrCounter + num) === word.length);
};

var isNumeric = function(c) {
    return (c.charCodeAt(0) >= '0'.charCodeAt(0)) && (c.charCodeAt(0) <= '9'.charCodeAt(0));
};

var getNumber = function(c) {
    return parseInt(c);
};
